### PR TITLE
fix: listpayments --count_total_payments should honor --creation_date_start and --creation_date_end

### PR DESCRIPTION
**Pull request by violetguos**
## Change Description
Fixes https://github.com/lightningnetwork/lnd/issues/8530

Add extra check for the creation start and end date
Only query all payments if those are not specified.

Added check for `totalCount` in existing test cases where the `CountTotal` flag isn't set.

## Steps to Test
Run unit tests and integration tests

### Testing
- [ ] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

## Summary by CodeRabbit

- **New Features**
	- Enhanced payment query capabilities to include filtering by creation date range.

- **Tests**
	- Added test for querying payments with specified start and end creation dates.

### CODE REVIEW

**Comment by violetguos** [comment]
@coderabbitai review

File: channeldb/payments.go [hunk 1]
```diff
@@ -596,7 +596,8 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
 		// Counting the total number of payments is expensive, since we
 		// literally have to traverse the cursor linearly, which can
 		// take quite a while. So it's an optional query parameter.
-		if query.CountTotal {
+		if query.CountTotal && query.CreationDateStart != 0 &&
```
Comment (violetguos) [inline channeldb/payments.go:599]: @coderabbitai generate unit tests for this if block

Comment (violetguos) [reply channeldb/payments.go:599]: Case 2 and 3 are covered by existing tests in `channeldb/payments_test.go`

Comment (violetguos) [reply channeldb/payments.go:599]: According to the original issue description, `CountTotal` should only be executed if we do not supply a creation start or end date. that means if we have `CreationDateStart` and `CreationDateEnd`, we want to count that transactions between these two dates, which is already done in the query handling block before we execute `CountTotal`. so `CountTotal` should not have the precedence if I understood the issue reporter's intent correctly

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Thanks for the PR! Could you fix the commit messages as required by our contribution guidelines?

**Comment by violetguos** [comment]
@yyforyongyu I have address your comment. could you please review?

**Comment by violetguos** [comment]
@ellemouton could you please review the changes? thanks

**Comment by ellemouton** [comment]
@violetguos  - it is in my review queue. Please have patience :)

File: channeldb/payments.go [hunk 2]
```diff
@@ -596,7 +596,10 @@ func (d *DB) QueryPayments(query PaymentsQuery) (PaymentsResponse, error) {
 		// Counting the total number of payments is expensive, since we
 		// literally have to traverse the cursor linearly, which can
 		// take quite a while. So it's an optional query parameter.
-		if query.CountTotal {
+		if query.CountTotal && query.CreationDateStart == 0 &&
+			query.CreationDateEnd == 0 {
+			// It will only take effect if CreationDateStart and
+			// CreationDateEnd are not specified.
 			var (
```
Comment (ellemouton) [inline channeldb/payments.go:603]: I dont think this is the behaviour we want. I think that we still want to count how many payments there were _but_ we just want to count the ones that happened between the given start and end times.

**Review by ellemouton — CHANGES_REQUESTED** [review]
Thanks for the PR @violetguos 🙏 

I dont think this is the behaviour we want though. If `count_total` is set at the same time as start and end dates, then we should count the total number of payments between those start and end dates. 

Other comments:
- please remove the `[skip-cl]` in the title of the last commit. We definitely want the CI to run here. 
- If you change a line of code in one commit - then it should be formatted correctly in that commit. Ie, dont add another commit to fix the formatting.

File: channeldb/payments.go (hunk 2 above)
Comment (violetguos) [reply channeldb/payments.go:603]: thanks for the feedback @ellemouton 

the original bug report by @andyschroder specifies that 
```
$ lncli listpayments --creation_date_start 1650000000 --count_total_payments --creation_date_end 1650000001|grep total_num_payments
    "total_num_payments": "0"
$
```
it seems like the original issue had a different interpretation on the expected behaviour

**Review by ellemouton — CHANGES_REQUESTED** [review]
see my reply to your comment & also see the previous review for comments about commit structuring :)

File: docs/release-notes/release-notes-0.18.0.md [hunk 3]
```diff
@@ -21,6 +21,11 @@
 - [Contributors (Alphabetical Order)](#contributors-alphabetical-order)
 
 # Bug Fixes
+* [Fix a bug](https://github.com/lightningnetwork/lnd/pull/8565) where 
+  `listpayments --count_total_payments` flag disregarded `--creation_date_start`
+  and `--creation_date_end`. The payments between the dates would be queried
+  instead of the total payments if the dates were supplied.
```
Comment (AndySchroder) [inline docs/release-notes/release-notes-0.18.0.md:27]: No, this is not correct.

File: channeldb/payments.go (hunk 2 above)
Comment (AndySchroder) [reply channeldb/payments.go:603]: yes, that time difference was specified to assume there were no payments between that short period of time.

Comment (AndySchroder) [reply channeldb/payments.go:603]: > I dont think this is the behaviour we want. I think that we still want to count how many payments there were _but_ we just want to count the ones that happened between the given start and end times.

yes, this was the original purpose of the bug

Comment (AndySchroder) [reply channeldb/payments.go:603]: > thanks for the feedback @ellemouton
> 
> the original bug report by @AndySchroder specifies that
> 
> ```
> $ lncli listpayments --creation_date_start 1650000000 --count_total_payments --creation_date_end 1650000001|grep total_num_payments
>     "total_num_payments": "0"
> $
> ```
> 
> it seems like the original issue had a different interpretation on the expected behaviour

no, please understand the interpretation that @ellemouton is providing.

**Comment by violetguos** [comment]
Sorry for the misinterpretation of the original example provided in the bug report. I currently don't have the capacity to implement the fix due to the change of scope.

**Pull request closed without merging**